sdks/v4-sdk/src/entities/pool.ts

```diff
@@ -0,0 +1,309 @@
+    ;[this.currency0, this.currency1] = sortsBefore(currencyA, currencyB)
+      ? [currencyA, currencyB]
+      : [currencyB, currencyA]
+    this.fee = fee
+    this.sqrtRatioX96 = JSBI.BigInt(sqrtRatioX96)
+    this.tickSpacing = tickSpacing
+    this.hooks = hooks
+    this.liquidity = JSBI.BigInt(liquidity)
+    this.tickCurrent = tickCurrent
+    this.tickDataProvider = Array.isArray(ticks) ? new TickListDataProvider(ticks, tickSpacing) : ticks
+    this.poolKey = Pool.getPoolKey(this.currency0, this.currency1, this.fee, this.tickSpacing, this.hooks)
+    this.poolId = Pool.getPoolId(this.currency0, this.currency1, this.fee, this.tickSpacing, this.hooks)
+
+    // define backwards compatibility interface
+    this.token0 = this.currency0
+    this.token1 = this.currency1
+    this.involvesToken = this.involvesCurrency
+    this.token0Price = this.currency0Price
+    this.token1Price = this.currency1Price
+  }
+
+  /**
+   * Returns true if the currency is either currency0 or currency1
+   * @param currency The currency to check
+   * @returns True if currency is either currency0 or currency
+   */
+  public involvesCurrency(currency: Currency): boolean {
+    return currency.equals(this.currency0) || currency.equals(this.currency1)
+  }
+
+  /**
+   * Returns the current mid price of the pool in terms of currency0, i.e. the ratio of currency1 over currency0
+   */
+  public get currency0Price(): Price<Currency, Currency> {
+    return (
+      this._currency0Price ??
+      (this._currency0Price = new Price(
+        this.currency0,
+        this.currency1,
+        Q192,
+        JSBI.multiply(this.sqrtRatioX96, this.sqrtRatioX96)
+      ))
+    )
+  }
+
+  /**
+   * Returns the current mid price of the pool in terms of currency1, i.e. the ratio of currency0 over currency1
+   */
+  public get currency1Price(): Price<Currency, Currency> {
+    return (
+      this._currency1Price ??
+      (this._currency1Price = new Price(
+        this.currency1,
+        this.currency0,
+        JSBI.multiply(this.sqrtRatioX96, this.sqrtRatioX96),
+        Q192
+      ))
+    )
+  }
+
+  /**
+   * Return the price of the given currency in terms of the other currency in the pool.
+   * @param currency The currency to return price of
+   * @returns The price of the given currency, in terms of the other.
+   */
+  public priceOf(currency: Currency): Price<Currency, Currency> {
+    invariant(this.involvesCurrency(currency), 'CURRENCY')
+    return currency.equals(this.currency0) ? this.currency0Price : this.currency1Price
+  }
+
+  /**
+   * Returns the chain ID of the currencies in the pool.
+   */
+  public get chainId(): number {
+    return this.currency0.chainId
+  }
+
+  public async getOutputAmount(
+    inputAmount: CurrencyAmount<Currency>,
+    sqrtPriceLimitX96?: JSBI
+  ): Promise<[CurrencyAmount<Currency>, Pool]> {
+    invariant(this.involvesCurrency(inputAmount.currency), 'CURRENCY')
```

Comment:
do we want something like "if the hook has beforeSwap or afterSwap, throw an error that we can't be sure"?

Reply:
ah nvm I see you have that in swap. Maybe a comment that clarifies this only gets output amount of the vanilla v3 swap piece?